encrypt/encrypt.go

```diff
@@ -27,14 +36,271 @@ var ArgHelp = `
+		os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ChecksumFileUnencMd5.Close()
+
+	ChecksumFileUnencSha256, err := os.OpenFile("checksum_unencrypted.sha256",
+		os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ChecksumFileUnencSha256.Close()
+
+	ChecksumFileEncMd5, err := os.OpenFile("checksum_encrypted.md5",
+		os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ChecksumFileEncMd5.Close()
+
+	ChecksumFileEncSha256, err := os.OpenFile("checksum_encrypted.sha256",
+		os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ChecksumFileEncSha256.Close()
+
+	// encrypt the input files
+	num_files := len(files)
+	for i, file := range files {
+		log.Infof("Encrypting file %v/%v: %s", i+1, num_files, file.unencrypted)
+
+		// encrypt the file
+		err = encrypt(file.unencrypted, file.encrypted, *publicKey, *privateKey)
+		if err != nil {
+			log.Fatal(err)
+		}
+		// calculate hashes
+		hashes, err := calculateHashes(file)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// Write hashes
+		ChecksumFileUnencMd5.WriteString(
+			fmt.Sprintf("%s %s\n", hashes.unencryptedMd5, file.unencrypted))
+		ChecksumFileUnencSha256.WriteString(
+			fmt.Sprintf("%s %s\n", hashes.unencryptedSha256, file.unencrypted))
+		ChecksumFileEncMd5.WriteString(
+			fmt.Sprintf("%s %s\n", hashes.encryptedMd5, file.encrypted))
+		ChecksumFileEncSha256.WriteString(
+			fmt.Sprintf("%s %s\n", hashes.encryptedSha256, file.encrypted))
+	}
+}
+
+// Checks that all the input files exists, and are readable
+func checkFiles(files []encryptionFileSet) error {
+	log.Info("Checking files")
+	for _, file := range files {
+		// check that the input file exists
+		fileInfo, err := os.Stat(file.unencrypted)
+		if err != nil {
+			return err
+		}
+		// check that the input file isn't a directory
+		if fileInfo.IsDir() {
+			return fmt.Errorf("%s is a directory", file.unencrypted)
+		}
+		// check that the output file doesn't exist
+		_, err = os.Stat(file.encrypted)
+		if err == nil {
+			return fmt.Errorf("outfile %s already exists", file.encrypted)
+		}
+	}
+	return nil
+}
+
+// Calculates md5 and sha256 hashes for the unencrypted and encrypted file
+func calculateHashes(fileSet encryptionFileSet) (*hashSet, error) {
+
+	hashes := hashSet{"", "", "", ""}
+
+	// open infile
+	inFile, err := os.Open(fileSet.unencrypted)
+	if err != nil {
+		return nil, err
+	}
+	defer inFile.Close()
+
+	// unencrypted md5
+	md5Hash := md5.New()
+	_, err = io.Copy(md5Hash, inFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.unencryptedMd5 = hex.EncodeToString(md5Hash.Sum(nil))
+
+	// reset file
+	_, err = inFile.Seek(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// unencrypted sha256
+	shaHash := sha256.New()
+	_, err = io.Copy(shaHash, inFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.unencryptedSha256 = hex.EncodeToString(shaHash.Sum(nil))
```

Comment:
This can be reworked a bit using the `TeeReader` or `MultiReader`, then you only need to read the file once while calculating both checksums.